### PR TITLE
Add ability to configure nodeSelector labels for k8s configs

### PR DESCRIPTION
**Pull request by ryan109**
## Purpose of PR
[Allows you to pin deployments to certain nodes based on nodeSelector labels](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes/#create-a-pod-that-gets-scheduled-to-your-chosen-node).

Note: this **only** allows for pinning deployments using nodeSelectors. [Other deployment pinning techniques](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/) such as affinity, anti-affinity, nodename etc have not been implemented

## Todos
Test nodeSelector works for
- [x] deployments
- [x] jobs
- [x] cronjobs

Tests have been completed against a repo only maintainers can see, it is referenced below in the PR log.

## Related links

### CODE REVIEW

File: kubetools/kubernetes/config/cronjob.py [hunk 1]
```diff
@@ -54,6 +55,12 @@ def make_cronjob_config(
         'containers': kubernetes_containers,
     }
 
+    if node_selector_labels is not None:
+        selector_labels = {}
+        for label, value in node_selector_labels.items():
+            selector_labels[label] = value
```
Comment (gchazot) [inline kubetools/kubernetes/config/cronjob.py]: Ugh, is this not `selector_labels = node_selector_labels.copy()` ?
I think we can even get away with not even using a copy and just referencing the input one. Given the structure of this, I don't see any reason why we would be trying to modify it down the line anyway

File: kubetools/kubernetes/config/deployment.py [hunk 2]
```diff
@@ -37,6 +38,12 @@ def make_deployment_config(
         'containers': kubernetes_containers,
     }
 
+    if node_selector_labels is not None:
+        selector_labels = {}
+        for label, value in node_selector_labels.items():
+            selector_labels[label] = value
+        template_spec['nodeSelector'] = selector_labels
```
Comment (gchazot) [inline kubetools/kubernetes/config/deployment.py]: same here

File: kubetools/kubernetes/config/job.py [hunk 3]
```diff
@@ -84,6 +85,12 @@ def make_job_config(
         'containers': [container],
     }
 
+    if node_selector_labels is not None:
+        selector_labels = {}
+        for label, value in node_selector_labels.items():
+            selector_labels[label] = value
+        template_spec['nodeSelector'] = selector_labels
```
Comment (gchazot) [inline kubetools/kubernetes/config/job.py]: same here

File: tests/configs/k8s_with_node_selectors/k8s_cronjobs.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            env:
+            - {name: KUBE, value: 'true'}
+            image: generic-image
+            imagePullPolicy: 'Always'
+            name: generic-container
+          restartPolicy: OnFailure
+          nodeSelector:
+          - nodeLabel: generic-label-value
+            nodeLabelName: generic-node-name 
+
```
Comment (gchazot) [inline tests/configs/k8s_with_node_selectors/k8s_cronjobs.yml]: Remove rogue whitespace
```suggestion

```

File: tests/configs/k8s_with_node_selectors/kubetools.yml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  generic-context:
+    image: generic-image
+    command: [generic-command]
+
+upgrades:
+  - name: Upgrade the database
+    containerContext: generic-context
+    nodeSelector:
+      nodeLabel: generic-label-value
+      nodeLabelName: generic-node-name
```
Comment (gchazot) [inline tests/configs/k8s_with_node_selectors/kubetools.yml]: These need to be more relevant, and less  confusing.

These tests are also here to document the capabilities of `kubetools` and, without going through k8s documentation, this makes it look like here we're defining 1 node selection label  with key `generic-node-name` and value `generic-node-value`...

File: tests/configs/k8s_with_node_selectors/kubetools.yml [hunk 6]
```diff
@@ -0,0 +1,44 @@
+name: generic-app-with-node-selectors
```
Comment (gchazot) [inline tests/configs/k8s_with_node_selectors/kubetools.yml:1]: Please reconsider the use of `generic-` here... The word does not seem to bring much descriptive value to the test.
I suspect that the word is coming from layers of copy-pasta of tests but originated from a test where it had some form of meaning (i.e. "generic" as opposed to another part of the test that would have been "specific"), and it doesn't seem to me to be the case here.

**Review by gchazot — CHANGES_REQUESTED** [review]
This seems structurally sound but needs some polishing

File: kubetools/kubernetes/config/cronjob.py (hunk 1 above)
Comment (ryan109) [reply kubetools/kubernetes/config/cronjob.py]: Ever been so tired you wrote code to rebuild a dict exactly as it was before?

Comment (ryan109) [reply kubetools/kubernetes/config/cronjob.py]: [7190f83](https://github.com/EDITD/kubetools/pull/161/commits/7190f83f41e4879c1d77975f3e8904be7f57c0d2)

File: kubetools/kubernetes/config/deployment.py (hunk 2 above)
Comment (ryan109) [reply kubetools/kubernetes/config/deployment.py]: [7190f83](https://github.com/EDITD/kubetools/pull/161/commits/7190f83f41e4879c1d77975f3e8904be7f57c0d2)

File: kubetools/kubernetes/config/job.py (hunk 3 above)
Comment (ryan109) [reply kubetools/kubernetes/config/job.py]: [7190f83](https://github.com/EDITD/kubetools/pull/161/commits/7190f83f41e4879c1d77975f3e8904be7f57c0d2)

File: tests/configs/k8s_with_node_selectors/k8s_cronjobs.yml (hunk 4 above)
Comment (ryan109) [reply tests/configs/k8s_with_node_selectors/k8s_cronjobs.yml]: [2885b4c](https://github.com/EDITD/kubetools/pull/161/commits/2885b4c368dd4e58c2b4de1416f8d0dd3166cec2)

File: tests/configs/k8s_with_node_selectors/kubetools.yml (hunk 5 above)
Comment (ryan109) [reply tests/configs/k8s_with_node_selectors/kubetools.yml]: [2885b4c](https://github.com/EDITD/kubetools/pull/161/commits/2885b4c368dd4e58c2b4de1416f8d0dd3166cec2)

File: tests/configs/k8s_with_node_selectors/kubetools.yml (hunk 6 above)
Comment (ryan109) [reply tests/configs/k8s_with_node_selectors/kubetools.yml:1]: [2885b4c](https://github.com/EDITD/kubetools/pull/161/commits/2885b4c368dd4e58c2b4de1416f8d0dd3166cec2)

**Review by gchazot — APPROVED** [review]

**Pull request merged**